File: hostel_management_system_backend/Controllers/UniversitiesController.cs
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("api/[controller]")]
public sealed class UniversitiesController : ControllerBase
{
    private readonly IUniversitiesService _service;

    public UniversitiesController(IUniversitiesService service)
    {
        _service = service;
    }

    [HttpGet]
    public async Task<ActionResult<List<UniversityReadDto>>> GetAll(CancellationToken cancellationToken)
    {
        var universities = await _service.GetAllAsync(cancellationToken);
        return Ok(universities);
    }

    [HttpGet("{id:guid}")]
    public async Task<ActionResult<UniversityReadDto>> GetById(Guid id, CancellationToken cancellationToken)
    {
        var university = await _service.GetByIdAsync(id, cancellationToken);
        return university is null ? NotFound() : Ok(university);
    }

    [HttpPost]
    public async Task<ActionResult<UniversityReadDto>> Create([FromBody] UniversityCreateDto dto, CancellationToken cancellationToken)
    {
        var created = await _service.CreateAsync(dto, cancellationToken);
        return CreatedAtAction(nameof(GetById), new { id = created.Id }, created);
    }

    [HttpPut("{id:guid}")]
    public async Task<ActionResult<UniversityReadDto>> Update(Guid id, [FromBody] UniversityUpdateDto dto, CancellationToken cancellationToken)
    {
        var updated = await _service.UpdateAsync(id, dto, cancellationToken);
        return updated is null ? NotFound() : Ok(updated);
    }

    [HttpDelete("{id:guid}")]
    public async Task<IActionResult> Delete(Guid id, CancellationToken cancellationToken)
    {
        var deleted = await _service.DeleteAsync(id, cancellationToken);
        return deleted ? NoContent() : NotFound();
    }
}

File: src/universities/universities.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
} from '@nestjs/common';

import { UniversityCreateDto, UniversityReadDto, UniversityUpdateDto } from './dto/universities.dto';
import { UniversitiesService } from './universities.service';

@Controller('universities')
export class UniversitiesController {
  constructor(private readonly universitiesService: UniversitiesService) {}

  @Get()
  getAll(): Promise<UniversityReadDto[]> {
    return this.universitiesService.getAll();
  }

  @Get(':id')
  getById(@Param('id', ParseUUIDPipe) id: string): Promise<UniversityReadDto> {
    return this.universitiesService.getById(id);
  }

  @Post()
  create(@Body() dto: UniversityCreateDto): Promise<UniversityReadDto> {
    return this.universitiesService.create(dto);
  }

  @Put(':id')
  update(
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: UniversityUpdateDto,
  ): Promise<UniversityReadDto> {
    return this.universitiesService.update(id, dto);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param('id', ParseUUIDPipe) id: string) {
    await this.universitiesService.delete(id);
  }
}
